src/pop_protocol.erl

```diff
@@ -363,3 +376,64 @@ generate_new_block(VerifierIndex, PD)
     check_block_map_structure(B2),
 
     B2.
+
+%% @doc Finds block by id.
+%% 
+%% Only looks among non-orphan blocks.
+%% Returns {ok, Block} or error.
+
+find_block_by_id(Id, PD)
+  when is_record(PD, protocol_data) ->
+    maps:find(Id, PD#protocol_data.tree_data#tree_data.block_map).
+
+
+%% @doc Adds block to the structure.
+%% 
+%% Block can be a duplicate, in which case it is ignored,
+%% or it can be an orphan, in which case it is saved
+%% to be added later.
+
+
+add_block(Block, CurrentTime, ProtocolData)
+  when 
+      is_record(ProtocolData, protocol_data), 
+      is_map(Block)
+      ->
+
+    %% checks maps for block and transactions inside
+    check_block_map_structure(Block),
+
+    OrphanMap = ProtocolData#protocol_data.orphan_blocks,
+    ThisId = maps:get(this_id, Block),
+    PrevId = maps:get(previous_id, Block),
+
+    R1 = find_block_by_id(ThisId, ProtocolData),
+    case R1 of
+	{ok, _} ->
+	    {ignored_duplicate, ProtocolData};
+	error ->
+	    R2 = find_block_by_id(maps:get(PrevId, Block), ProtocolData),
+	    case R2 of
+		{ok, _} ->
+		    {added_new, add_block_in_order(Block, CurrentTime, ProtocolData)};
+		error ->
+		    OrphanMapNew = maps:put(ThisId, Block, OrphanMap),
+		    {orphan, ProtocolData#protocol_data{orphan_blocks = OrphanMapNew}}
```

Comment:
nice!

Reply:
I think "orphan" is actually inconsistent with the usage of the term in bitcoin.

Here orphan means there is no parents

in bitcoin, all orphan blocks have parents but are not part of the longest chain

😱

Reply:
`all orphan blocks have parents`

ಠ_ಠ

Reply:
:o?

Reply:
I guess it's like they all got abandoned by their parent in the longest chain 😱😱😱😱